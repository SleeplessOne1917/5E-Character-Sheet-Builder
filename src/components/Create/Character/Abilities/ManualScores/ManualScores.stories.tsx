import { ComponentMeta, ComponentStory } from '@storybook/react';
import ManualScores, { ManualScoresProps } from './ManualScores';
import MockStore, { mockAbilities } from '../MockAbilitiesStore';

export default {
	title: 'Components/CharacterCreation/Abilities/ManualScores',
	component: ManualScores,
	args: {
		abilities: mockAbilities
	}
} as ComponentMeta<typeof ManualScores>;

export const Default: ComponentStory<typeof ManualScores> = (
	args: ManualScoresProps
) => <ManualScores {...args} />;
Default.decorators = [story => <MockStore>{story()}</MockStore>];
