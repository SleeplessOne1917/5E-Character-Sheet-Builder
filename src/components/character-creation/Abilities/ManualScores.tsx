import { ChangeEvent, FocusEvent, useCallback, useState } from 'react';
import { useAppDispatch, useAppSelector } from '../../../hooks/reduxHooks';

import { AbilityItem } from '../../../types/srd';
import AbilityScores from '../../../types/abilityScores';
import classes from './ManualScores.module.css';
import { getTotalScore } from '../../../services/abilityScoreService';
import { updateBase } from '../../../redux/features/abilityScores';
import useGetAbilityScore from '../../../hooks/useGetAbilityScore';

type ManualScoresProps = {
	abilities: AbilityItem[];
};

type AbilityScoreStrings = {
	str?: string | null;
	dex?: string | null;
	con?: string | null;
	int?: string | null;
	wis?: string | null;
	cha?: string | null;
};

const ManualScores = ({ abilities }: ManualScoresProps): JSX.Element => {
	const dispatch = useAppDispatch();
	const getAbilityScore = useGetAbilityScore();
	const [abilityScoreStrings, setAbilityScoreStrings] =
		useState<AbilityScoreStrings>({});

	const handleChange = useCallback(
		(event: ChangeEvent<HTMLInputElement>, abilityIndex: AbilityScores) => {
			setAbilityScoreStrings(prevState => ({
				...prevState,
				[abilityIndex]: event.target.value
			}));
		},
		[setAbilityScoreStrings]
	);

	const handleBlur = useCallback(
		(event: FocusEvent<HTMLInputElement>, abilityIndex: AbilityScores) => {
			let value: number | null = parseInt(event.target.value);
			if (isNaN(value)) {
				value = null;
			} else {
				if (value < 3) {
					value = 3;
				}

				if (value > 18) {
					value = 18;
				}
			}

			setAbilityScoreStrings(prevState => ({
				...prevState,
				[abilityIndex]: value
			}));
			dispatch(updateBase({ value, abilityIndex }));
		},
		[dispatch, setAbilityScoreStrings]
	);

	return (
		<div className={classes['manual-scores']}>
			{abilities.map(ability => (
				<div key={ability.index} className={classes['manual-score']}>
					<label htmlFor={`manual-${ability.index}`}>{ability.full_name}</label>
					<input
						id={`manual-${ability.index}`}
						type="text"
						onBlur={event => handleBlur(event, ability.index as AbilityScores)}
						value={
							(abilityScoreStrings[ability.index as AbilityScores]
								? abilityScoreStrings[ability.index as AbilityScores]
								: '') as string
						}
						onChange={event =>
							handleChange(event, ability.index as AbilityScores)
						}
					/>
					<div className={classes.total}>
						Total:{' '}
						{getAbilityScore(ability.index as AbilityScores).base
							? getTotalScore(getAbilityScore(ability.index as AbilityScores))
							: '\u2014'}
					</div>
				</div>
			))}
		</div>
	);
};

export default ManualScores;
