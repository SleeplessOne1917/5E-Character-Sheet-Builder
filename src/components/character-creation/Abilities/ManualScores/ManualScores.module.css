.manual-scores {
    display: flex;
    flex-wrap: wrap;
    align-self: center;
}

.manual-score {
    width: 28%;
    color: var(--davy);
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 0 0.5rem;
}

.manual-score input {
    width: 100%;
    font-size: 1.5rem;
    font-family: inherit;
    border: 1px solid var(--davy);
    background-color: var(--heaven-light);
    text-align: center;
}

.manual-score input:focus {
    outline: var(--focus-outline);
}

.manual-score label {
    font-size: 1.3rem;
    font-weight: bold;
    margin-bottom: 0.5rem;
}

.total {
    font-weight: bold;
    margin-top: 1rem;
}

@media screen and (min-width: 1024px) {
    .manual-score {
        width: 14%;
    }
}