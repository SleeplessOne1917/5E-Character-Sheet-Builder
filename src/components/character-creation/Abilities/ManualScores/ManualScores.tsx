import {
	AbilityScoresState,
	updateBase
} from '../../../../redux/features/abilityScores';
import { ChangeEvent, FocusEvent, useCallback, useState } from 'react';
import { useAppDispatch, useAppSelector } from '../../../../hooks/reduxHooks';

import { AbilityItem } from '../../../../types/srd';
import AbilityScores from '../../../../types/abilityScores';
import classes from './ManualScores.module.css';
import { getTotalScore } from '../../../../services/abilityScoreService';
import useGetAbilityScore from '../../../../hooks/useGetAbilityScore';

export type ManualScoresProps = {
	abilities: AbilityItem[];
};

type AbilityScoreStrings = {
	str?: string | null;
	dex?: string | null;
	con?: string | null;
	int?: string | null;
	wis?: string | null;
	cha?: string | null;
};

const getInitialAbilityScoreStrings = (abilityScores: AbilityScoresState) =>
	Object.entries(abilityScores).reduce(
		(prev, [key, value]) => ({
			...prev,
			[key]: value.base ? `${value.base}` : null
		}),
		{}
	);

const ManualScores = ({ abilities }: ManualScoresProps): JSX.Element => {
	const abilityScores = useAppSelector(
		state => state.editingCharacter.abilityScores
	);
	const dispatch = useAppDispatch();
	const getAbilityScore = useGetAbilityScore();
	const [abilityScoreStrings, setAbilityScoreStrings] =
		useState<AbilityScoreStrings>(getInitialAbilityScoreStrings(abilityScores));

	const handleChange = useCallback(
		(event: ChangeEvent<HTMLInputElement>, abilityIndex: AbilityScores) => {
			setAbilityScoreStrings(prevState => ({
				...prevState,
				[abilityIndex]: event.target.value
			}));
		},
		[setAbilityScoreStrings]
	);

	const handleBlur = useCallback(
		(event: FocusEvent<HTMLInputElement>, abilityIndex: AbilityScores) => {
			let value: number | null = parseInt(event.target.value);
			if (isNaN(value)) {
				value = null;
			} else {
				if (value < 3) {
					value = 3;
				}

				if (value > 18) {
					value = 18;
				}
			}

			setAbilityScoreStrings(prevState => ({
				...prevState,
				[abilityIndex]: value
			}));
			dispatch(updateBase({ value, abilityIndex }));
		},
		[dispatch, setAbilityScoreStrings]
	);

	return (
		<div
			className={classes['manual-scores']}
			role="region"
			aria-label="Manual Scores"
		>
			{abilities.map(ability => (
				<div key={ability.index} className={classes['manual-score']}>
					<label htmlFor={`manual-${ability.index}`}>{ability.full_name}</label>
					<input
						id={`manual-${ability.index}`}
						type="text"
						onBlur={event => handleBlur(event, ability.index as AbilityScores)}
						value={
							(abilityScoreStrings[ability.index as AbilityScores]
								? abilityScoreStrings[ability.index as AbilityScores]
								: '') as string
						}
						onChange={event =>
							handleChange(event, ability.index as AbilityScores)
						}
						placeholder="&mdash;"
					/>
					<div className={classes.total}>
						Total:{' '}
						{getAbilityScore(ability.index as AbilityScores).base
							? getTotalScore(getAbilityScore(ability.index as AbilityScores))
							: '\u2014'}
					</div>
				</div>
			))}
		</div>
	);
};

export default ManualScores;
