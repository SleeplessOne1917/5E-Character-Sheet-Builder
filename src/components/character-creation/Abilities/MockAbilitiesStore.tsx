import { configureStore, createSlice } from '@reduxjs/toolkit';

import { AbilityScore } from '../../../redux/features/abilityScores';
import AbilityScores from '../../../types/abilityScores';
import { Provider } from 'react-redux';
import { ReactNode } from 'react';

const defaultInitialState: { abilityScores: AbilityScoresState } = {
	abilityScores: {
		str: {
			base: 12
		},
		dex: {
			base: 8,
			abilityImprovement: 4
		},
		cha: { base: 15, raceBonus: 2 },
		con: { override: 10 },
		int: { otherBonus: 5 },
		wis: { base: 14, override: 25 }
	}
};

const getMockStore = (index: AbilityScores, overrideValues: AbilityScore) => {
	const initialState = {
		abilityScores: {
			...defaultInitialState.abilityScores,
			[index]: {
				...defaultInitialState.abilityScores[index],
				...overrideValues
			}
		}
	};

	const mockEditingCharacterSlice = createSlice({
		name: 'editingCharacter',
		initialState: initialState,
		reducers: {}
	});

	return configureStore({
		reducer: { editingCharacter: mockEditingCharacterSlice.reducer }
	});
};

const MockStore = ({
	overrideValues = {},
	index,
	children
}: {
	overrideValues?: AbilityScore;
	index: AbilityScores;
	children: ReactNode;
}) => (
	<Provider store={getMockStore(index, overrideValues)}>{children}</Provider>
);

export default MockStore;

export const mockAbilities = [
	{ index: 'str', full_name: 'Strength' },
	{ index: 'dex', full_name: 'Dexterity' },
	{ index: 'con', full_name: 'Constitution' },
	{ index: 'int', full_name: 'Intelligence' },
	{ index: 'wis', full_name: 'Wisdom' },
	{ index: 'cha', full_name: 'Charisma' }
];
