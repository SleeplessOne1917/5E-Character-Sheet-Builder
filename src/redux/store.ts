import {
	AnyAction,
	ConfigureStoreOptions,
	ListenerEffect,
	Reducer,
	ThunkDispatch,
	configureStore,
	createListenerMiddleware
} from '@reduxjs/toolkit';
import editingCharacter, {
	EditingCharacterState
} from './features/editingCharacter';
import { get, set } from 'idb-keyval';
import { useEffect, useState } from 'react';

import generationMethod from './features/generationMethod';
import rollGroups from './features/rollGroups';
import toast from './features/toast';
import viewer from './features/viewer';

const getReducer = () => ({
	toast,
	viewer,
	editingCharacter: editingCharacter as Reducer<
		EditingCharacterState,
		AnyAction
	>,
	rollGroups,
	generationMethod
});

const indexedDbCacheListener = createListenerMiddleware();

const REDUX_CACHE_KEY = 'redux-cache';

const updateCache: ListenerEffect<
	AnyAction,
	unknown,
	ThunkDispatch<unknown, unknown, AnyAction>,
	unknown
> = (action, { getState }) => {
	set(REDUX_CACHE_KEY, getState());
};

const updateCachePredicate = () => true;

const getStore = async () => {
	indexedDbCacheListener.startListening({
		predicate: updateCachePredicate,
		effect: updateCache
	});

	const options: ConfigureStoreOptions = {
		reducer: getReducer(),
		middleware: getDefaultMiddleware =>
			getDefaultMiddleware().prepend(indexedDbCacheListener.middleware)
	};

	const reduxCache = await get(REDUX_CACHE_KEY);
	if (reduxCache) {
		options.preloadedState = reduxCache;
	}

	const theStore = configureStore(options);
	if (!reduxCache) {
		set(REDUX_CACHE_KEY, theStore.getState());
	}

	return theStore;
};

export const useStore = () => {
	const [theStore, setTheStore] = useState(getTestStore());
	const [loadingStore, setLoadingStore] = useState(true);

	useEffect(() => {
		getStore().then(s => {
			setTheStore(s);
			setLoadingStore(false);
		});

		return () => {
			indexedDbCacheListener.stopListening({
				predicate: updateCachePredicate,
				effect: updateCache
			});
		};
	}, [setTheStore]);

	return { store: theStore, loading: loadingStore };
};

export const getTestStore = (
	reducerOverrides?: Partial<ReturnType<typeof getReducer>>
) =>
	configureStore({
		reducer: { ...getReducer(), ...reducerOverrides },
		devTools: false
	});

const store = getTestStore();

export type RootState = ReturnType<typeof store.getState>;
export type Dispatch = typeof store.dispatch;
