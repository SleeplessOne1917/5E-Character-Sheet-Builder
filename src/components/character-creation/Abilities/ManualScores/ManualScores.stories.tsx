import { ComponentMeta, ComponentStory } from '@storybook/react';
import ManualScores, { ManualScoresProps } from './ManualScores';
import MockStore, { mockAbilities } from '../MockAbilitiesStore';

const ManualScoresSb = {
	title: 'Components/ManualScores',
	component: ManualScores,
	args: {
		abilities: mockAbilities
	}
} as ComponentMeta<typeof ManualScores>;

export default ManualScoresSb;

export const Default: ComponentStory<typeof ManualScores> = (
	args: ManualScoresProps
) => <ManualScores {...args} />;
Default.decorators = [story => <MockStore index="dex">{story()}</MockStore>];
