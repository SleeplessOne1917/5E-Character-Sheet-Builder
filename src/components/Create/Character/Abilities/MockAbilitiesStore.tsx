import { AbilityScoresState } from '../../../../redux/features/abilityScores';
import { Provider } from 'react-redux';
import { ReactNode } from 'react';
import { AbilityItem } from '../../../../types/srd';
import { getTestStore } from '../../../../redux/store';
import getMockEditingCharacter from '../../../../mock/editingCharacterMock';

const defaultInitialState: AbilityScoresState = {
	str: {
		base: 12,
		highest: 20,
		abilityImprovement: 0
	},
	dex: {
		base: 8,
		abilityImprovement: 4,
		highest: 20
	},
	cha: { base: 15, raceBonus: 2, highest: 20, abilityImprovement: 0 },
	con: { override: 10, abilityImprovement: 0, highest: 20 },
	int: { otherBonus: 5, abilityImprovement: 0, highest: 20 },
	wis: { base: 14, override: 25, abilityImprovement: 0, highest: 20 }
};

const MockStore = ({
	overrideValues,
	children
}: {
	overrideValues?: Partial<AbilityScoresState>;
	children: ReactNode;
}) => (
	<Provider
		store={getTestStore({
			editingCharacter: getMockEditingCharacter({
				abilityScores: { ...defaultInitialState, ...overrideValues }
			})
		})}
	>
		{children}
	</Provider>
);

export default MockStore;

export const mockAbilities: AbilityItem[] = [
	{ index: 'str', full_name: 'Strength' },
	{ index: 'dex', full_name: 'Dexterity' },
	{ index: 'con', full_name: 'Constitution' },
	{ index: 'int', full_name: 'Intelligence' },
	{ index: 'wis', full_name: 'Wisdom' },
	{ index: 'cha', full_name: 'Charisma' }
];
