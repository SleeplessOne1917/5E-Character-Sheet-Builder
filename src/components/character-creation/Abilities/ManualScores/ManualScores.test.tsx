import '@testing-library/jest-dom';

import * as stories from './ManualScores.stories';

import { render, screen } from '@testing-library/react';

import ManualScores from './ManualScores';
import { Provider } from 'react-redux';
import { composeStories } from '@storybook/testing-react';
import { getTestStore } from '../../../../redux/store';
import { mockAbilities } from '../MockAbilitiesStore';
import userEvent from '@testing-library/user-event';

const { Default } = composeStories(stories);

it('renders correctly', () => {
	render(<Default />);

	expect(screen.getByRole('region')).toMatchSnapshot();
});

describe('text boxes', () => {
	it('should not allow non number input', async () => {
		render(<Default />);

		const inputs = screen.getAllByRole('textbox');
		for (const input of inputs) {
			await userEvent.clear(input);
			await userEvent.type(input, 'abc');
			await userEvent.click(screen.getAllByText(/Total/i).at(0) as HTMLElement);

			expect((input as HTMLInputElement).value).toBe('');
		}
	});

	it('should not allow input lower than 3', async () => {
		render(<Default />);

		const inputs = screen.getAllByRole('textbox');
		for (const input of inputs) {
			await userEvent.clear(input);
			await userEvent.type(input, '2');
			await userEvent.click(screen.getAllByText(/Total/i).at(0) as HTMLElement);

			expect((input as HTMLInputElement).value).toBe('3');
		}
	});

	it('should not allow input higher than 18', async () => {
		render(<Default />);

		const inputs = screen.getAllByRole('textbox');
		for (const input of inputs) {
			await userEvent.clear(input);
			await userEvent.type(input, '19');
			await userEvent.click(screen.getAllByText(/Total/i).at(0) as HTMLElement);

			expect((input as HTMLInputElement).value).toBe('18');
		}
	});
});

it('only shows total when base score is set', async () => {
	render(
		<Provider store={getTestStore()}>
			<ManualScores abilities={mockAbilities} />
		</Provider>
	);

	expect(screen.queryByText(/Total: \d/i)).not.toBeInTheDocument();

	const input = screen.getAllByRole('textbox').at(0) as HTMLInputElement;
	await userEvent.type(input, '14');
	await userEvent.click(screen.getAllByText(/Total/i).at(0) as HTMLElement);

	expect(screen.getByText(/Total: \d/i)).toBeInTheDocument();
});
