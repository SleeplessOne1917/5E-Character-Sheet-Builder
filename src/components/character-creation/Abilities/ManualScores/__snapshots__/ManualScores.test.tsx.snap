// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  aria-label="Manual Scores"
  class="manual-scores"
  role="region"
>
  <div
    class="manual-score"
  >
    <label
      for="manual-str"
    >
      Strength
    </label>
    <input
      id="manual-str"
      type="text"
      value="12"
    />
    <div
      class="total"
    >
      Total:
       
      12
    </div>
  </div>
  <div
    class="manual-score"
  >
    <label
      for="manual-dex"
    >
      Dexterity
    </label>
    <input
      id="manual-dex"
      type="text"
      value="8"
    />
    <div
      class="total"
    >
      Total:
       
      12
    </div>
  </div>
  <div
    class="manual-score"
  >
    <label
      for="manual-con"
    >
      Constitution
    </label>
    <input
      id="manual-con"
      type="text"
      value=""
    />
    <div
      class="total"
    >
      Total:
       
      —
    </div>
  </div>
  <div
    class="manual-score"
  >
    <label
      for="manual-int"
    >
      Intelligence
    </label>
    <input
      id="manual-int"
      type="text"
      value=""
    />
    <div
      class="total"
    >
      Total:
       
      —
    </div>
  </div>
  <div
    class="manual-score"
  >
    <label
      for="manual-wis"
    >
      Wisdom
    </label>
    <input
      id="manual-wis"
      type="text"
      value="14"
    />
    <div
      class="total"
    >
      Total:
       
      25
    </div>
  </div>
  <div
    class="manual-score"
  >
    <label
      for="manual-cha"
    >
      Charisma
    </label>
    <input
      id="manual-cha"
      type="text"
      value="15"
    />
    <div
      class="total"
    >
      Total:
       
      17
    </div>
  </div>
</div>
`;
